draft-ietf-quic-transport.md

```diff
@@ -881,33 +909,30 @@ Incoming packets are classified on receipt.  Packets can either be associated
 with an existing connection, or - for servers - potentially create a new
 connection.
 
-Hosts try to associate the packet with an existing connection. If the
-packet has a connection ID corresponding to an existing connection, QUIC
+Hosts try to associate the packet with an existing connection. If the packet has
+a Destination Connection ID corresponding to an existing connection, QUIC
 processes that packet accordingly. Note that a NEW_CONNECTION_ID frame
-({{frame-new-connection-id}}) would associate more than one connection ID
-with a connection.
+({{frame-new-connection-id}}) would associate more than one connection ID with a
+connection.
 
-If there is no connection ID, but the packet matches the address/port tuple
-of a connection where the host did not require connection IDs, QUIC processes
-the packet as part of that connection. Endpoints MUST drop packets that omit
-connection IDs if they do not meet both of these criteria.
+If the Destination Connection ID is zero length, but the packet matches the
+address/port tuple of a connection where the host did not require connection
+IDs, QUIC processes the packet as part of that connection. Endpoints MUST drop
+packets with zero-length Destination Connection ID fields if they do not meet
+both of these criteria.
```

Comment:
This adds a bit of complexity that I'm not sure is worth it. As I understand it, it would require a look up table of all remote address/port tuples in use, that would have to be checked for every new connection attempt. If we don't do this, and leave it up to the peer to make the smart choice, one of two things happen:

1) The do things correctly and everything works.
2) They have two connections on the same tuple, 1 with a CID and one without. The one without breaks down because it cannot distinguish its short header packets from the other one.

If the peer does 2) above, then it's their own fault and they suffer the consequences.

Reply:
With the `omit` setting removed, I think there is no more "did not require Connection IDs," only connections where the Connection ID is zero-length.  At that point, this is a duplicate of CID collision.

In that case, the language can be something like "If the Destination Connection ID matches multiple connections (such as because they all have zero-length connection IDs), the address/port tuple is used to identify the correct connection.  Endpoints MUST drop packets with zero-length connection IDs if they cannot be matched to a connection."

Reply:
For short headers, an endpoint can simplify its job by always using the same length CID, but if it doesn't then it probably needs to keep the connections in a tree by CID.  Check the first byte, see if there are any CIDs that begin with that byte, and so on, until it either finds a complete match or a lookup failure.

Then, if there's a lookup failure, it might have been zero-length and you were accidentally reading the packet number and encrypted payload as a CID.  So now you look into your separate bag of connections where the CID is zero-length and try to match by tuple.

But wait!  You also might have had a collision between the packet number and a CID, so packets which can't be validated by the selected connection *also* need to be checked against the zero-length set.  And we pray that doesn't happen very often, because this seems like the worst-case path through the code.

Ick.

Reply:
"don't do that"